lite/kernels/nnadapter/engine.cc

```diff
@@ -329,49 +339,55 @@ bool Engine::Run() {
   for (size_t i = 0; i < input_count; i++) {
     input_dims[i] = input_vars_[i].value->dims().Vectorize();
   }
-  // Find the compiled device program according to the input dimensions
-  std::shared_ptr<Program> program = nullptr;
-  if (!programs_.count(input_dims)) {
-    // Rebuild the device program corresponding to the input dimensions if not
-    // found
-    std::vector<std::string> device_names;
-    for (auto* device : devices_) {
-      const char* name = nullptr;
-      NNAdapterDevice_getName_invoke(device, &name);
-      device_names.push_back(name);
+  // try to execute all programs.
+  for (auto program : programs_) {
+    int ret = program->Execute();
+    if (ret == NNADAPTER_INVALID_DIMENSIONS) {
+      VLOG(1)
+          << "Input shapes are invalid for the program, try the next program.";
+      continue;
     }
-    program = std::make_shared<Program>(context_);
-    // Take the model cache buffer from the scope
-    std::vector<char> model_cache_buffer;
-    // Generate a cache token based on the input names and shapes
-    auto model_cache_token = GenerateModelCacheToken(device_names, input_vars_);
-    VLOG(3) << "NNAdapter model_cache_token: " << model_cache_token;
-    ctx_->As<NNAdapterContext>().NNAdapterModelCacheBuffers(
-        exec_scope_, model_cache_token, &model_cache_buffer);
-    VLOG(3) << "NNAdapter model_cache_buffer size: "
-            << model_cache_buffer.size();
-    // Load the compiled device program from the model cache buffer or file
-    if (!program->LoadFromCache(
-            model_cache_token, &model_cache_buffer, model_cache_dir_)) {
-      // Compile the model online to generate the device program and cache it to
-      // the file
-      CHECK(program->BuildAndCacheToFile(block_idx_,
-                                         program_desc_,
-                                         exec_scope_,
-                                         input_vars_,
-                                         &output_vars_,
-                                         model_cache_token,
-                                         model_cache_dir_));
-    }
-    CHECK(program->IsValid());
-    CHECK(program->SetInputsAndOutputs(&input_vars_, &output_vars_));
-    programs_[input_dims] = program;
-  } else {
-    program = programs_[input_dims];
-    CHECK(program);
-    CHECK(program->IsValid());
+    CHECK_EQ(ret, NNADAPTER_NO_ERROR) << "Program execute failed.";
+    return true;
+  }
+  // Rebuild the device program corresponding to the input dimensions if not
+  // find valid program.
+  VLOG(1) << "Not found valid program for the current input shapes. New "
```

Comment:
No suitable program found for current input shapes, try generating a new program online.

Reply:
done